xchange-lgo/src/main/java/info/bitrich/xchangestream/lgo/LgoStreamingTradeService.java

```diff
@@ -0,0 +1,287 @@
+package info.bitrich.xchangestream.lgo;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import info.bitrich.xchangestream.core.StreamingTradeService;
+import info.bitrich.xchangestream.lgo.domain.LgoMatchOrderEvent;
+import info.bitrich.xchangestream.lgo.domain.LgoOrderEvent;
+import info.bitrich.xchangestream.lgo.dto.*;
+import info.bitrich.xchangestream.service.netty.StreamingObjectMapperHelper;
+import io.reactivex.Observable;
+import org.knowm.xchange.currency.CurrencyPair;
+import org.knowm.xchange.dto.Order;
+import org.knowm.xchange.dto.trade.LimitOrder;
+import org.knowm.xchange.dto.trade.MarketOrder;
+import org.knowm.xchange.dto.trade.OpenOrders;
+import org.knowm.xchange.dto.trade.UserTrade;
+import org.knowm.xchange.lgo.LgoAdapters;
+import org.knowm.xchange.lgo.dto.key.LgoKey;
+import org.knowm.xchange.lgo.dto.order.LgoEncryptedOrder;
+import org.knowm.xchange.lgo.dto.order.LgoOrderSignature;
+import org.knowm.xchange.lgo.dto.order.LgoPlaceCancelOrder;
+import org.knowm.xchange.lgo.dto.order.LgoPlaceOrder;
+import org.knowm.xchange.lgo.service.CryptoUtils;
+import org.knowm.xchange.lgo.service.LgoKeyService;
+import org.knowm.xchange.lgo.service.LgoSignatureService;
+import si.mazi.rescu.SynchronizedValueFactory;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toMap;
+
+public class LgoStreamingTradeService implements StreamingTradeService {
+
+    private final LgoStreamingService streamingService;
+    private final LgoKeyService keyService;
+    private final LgoSignatureService signatureService;
+    private final SynchronizedValueFactory<Long> nonceFactory;
+    private final Map<CurrencyPair, Map<String, Order>> allOrders = new HashMap<>();
+    private final Map<CurrencyPair, Observable<LgoBatchUpdate>> orderBatchChangesSubscriptions = new HashMap<>();
+    private Observable<LgoOrderEvent> afrSubscription;
+
+    public LgoStreamingTradeService(LgoStreamingService streamingService, LgoKeyService keyService, LgoSignatureService signatureService, SynchronizedValueFactory<Long> nonceFactory) {
+        this.streamingService = streamingService;
+        this.keyService = keyService;
+        this.signatureService = signatureService;
+        this.nonceFactory = nonceFactory;
+    }
+
+    /**
+     * {@inheritDoc}
+     * First sent orders will be current open orders.
+     */
+    @Override
+    public Observable<Order> getOrderChanges(CurrencyPair currencyPair, Object... args) {
+        return getOrderBatchChanges(currencyPair)
+                .flatMap(Observable::fromIterable);
+    }
+
+    /**
+     * Get an up-to-date view of all your open orders after each LGO batch execution.
+     * First sending will be the actual open orders list.
+     */
+    public Observable<OpenOrders> getOpenOrders(CurrencyPair currencyPair) {
```

Comment:
In my experience, this is very unlikely to be stable.  There's no way for Netty to tell if a web socket has dropped and you have missed messages, unless the exchange provides a sequentially numbered heartbeat and  a clear guide about how to refresh from the REST API in such a way that you can safely discard websocket messages in the meantime. Binance has an excellent guide for how to do this for their order book API, but order books are updated constantly so there are plenty of sync opportunities. Open orders don't update very often (unless you're doing arbitrage).

I always write code that _both_ polls the REST API (on a longer timescale) _and_ monitors the websocket (to get faster updates) and manages the corresponding snapshot in a layer above both xchange-stream and XChange.  The regular polling ensures that if inconsistencies appear, they don't last long in memory.

This approach looks like it has no self-correction mechanism (although I can't be sure without having read more about LGO, so please correct me if I am wrong) in which case if this cache gets out of date, it will stay out of date until the open orders are changed while the web socket is properly connected.

When discussing this API, we spent quite a long time trying to decide whether xchange-stream should maintain a snapshot of the orders like this, make REST API calls for synchronisation and so on, but couldn't find any way of doing it reliably across exchanges so we avoided it.

Reply:
Interesting. 
LGO sends monotonic batch_id on each message, so a missed message could be detected. In that case, it's just a matter of unsubscribing/subscribing to this channel to receive a fresh snapshot. We could implement that. What do you think ? 
In certain conditions, LGO will also send a snapshot if it thinks some data may have been missed. That case is already dealt with.

Reply:
That does sound good, as long as the {{StreamingTradeService}} manages the disconnection/reconnection transparently, _without disrupting the existing `Observable`s_ (so the client isn't required to close everything and fetch new `Observable`s).  That would mean switching _all_ the exposed `Observable`s on all of the `Streaming...Service` implementations to `Subject`s.  That is a big change!

However, it still doesn't solve the whole problem. You've still got the situation where:

# Socket opens
# Initial snapshot is sent
# Websocket (silently) lost
# Update is sent
# Websocket (silently) recovers
# Time passes where no order updates are sent and the open orders are out of date
# Update is sent - missing message is detected and recovery occirs

What I found was that this happens surprisingly often, and given that open orders don't necessarily update very often, that time gap can be very large.

Reply:
Basically I've got no issue with whatever approach you want to use for this - my only concern is that if the feature appears in this library, it needs to be solid and work without the client having to do anything "special".  If we can come up with such a solution, fantastic. If not, it might be safer to omit it for the time being.

Reply:
I see you point, thank you for your feedback. We'll look at our options in the next few days. 

Thinking out loud : A traditional way in socket land to deal with connection loss detection is to send bi-directional keep-alive messages. I guess one could use the standard PING mechanism of websockets to implement that, to make sure that in the absence of messages after a given time, it's not due to connectivity issues.

Reply:
Yes indeed. In fact, IIRC, there's support in NettyStreamingService already from when I tried to do exactly this (and gave up!)

Reply:
OK it looks like you're getting full snapshots every time and aggressively keeping the socket alive, so this is probably going to work.  Fingers crossed! 👍